### PR TITLE
Add support for record iteration

### LINKED ISSUE #10080: Add support for iterating records
**Description:**
Currently, iteration is not supported for records. Need to add iteration support for records so that users can iterate through the fields of a record.

### PR DESCRIPTION
**Pull request by pubudu91**
### Foreach loop
```ballerina
Person p = { name: "John Doe", age: 25, address: { street: "Palm Grove", city: "Colombo 3" } };

foreach f, v in p {
    io:println(f + ":" + <string>v);
}
```
### Iterable operations
```ballerina
Foo f = {a: "AA", b: "BB", c: "CC", d: "DD", e: "EE"};

map<string> newf = f.map(((string, string) entry) => (string, string) {
                var (field, value) = entry;
                return (field, value.toLower());
            })
            .filter(((string, string) entry) => boolean {
                var (field, value) = entry;
                if (value != "aa" && value != "ee") {
                    return true;
                }
                return false;
            });
```

### CODE REVIEW

**Comment by vinok88** [comment]
The commit message of the commit 03b0050 (sep 4th) is too long. Please wrap the topic to 50 to 72 characters and put any other information in the body.

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticCode.java [hunk 1]
```diff
@@ -231,6 +231,7 @@
     ITERABLE_NOT_SUPPORTED_OPERATION("iterable.not.supported.operation"),
     ITERABLE_TOO_MANY_VARIABLES("iterable.too.many.variables"),
     ITERABLE_NOT_ENOUGH_VARIABLES("iterable.not.enough.variables"),
+    ITERABLE_REQUIRES_N_VARIABLES("iterable.requires.n.variables"),
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticCode.java]: Can't we reuse one of `ITERABLE_NOT_ENOUGH_VARIABLES` or `ITERABLE_TOO_MANY_VARIABLES`?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IterableAnalyzer.java [hunk 2]
```diff
@@ -155,20 +156,20 @@ private void handleLambdaBasedIterableOperation(IterableContext context, Operati
         final List<BType> givenArgTypes = calculatedGivenInputArgs(operation);
         final List<BType> givenRetTypes = calculatedGivenOutputArgs(operation);
 
-        final List<BType> expectedArgTypes = calculateExpectedInputArgs(operation);
-        final List<BType> expectedRetTypes = calculateExpectedOutputArgs(operation, givenRetTypes);
+        final List<BType> actualArgTypes = calculateExpectedInputArgs(operation);
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IterableAnalyzer.java]: the function name and the variable name doesn't seems to match.
Line 156 calculates the givenArgType (actuals), so shouldn't this be the `expectedArgTypes`?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IterableAnalyzer.java [hunk 3]
```diff
@@ -225,17 +226,18 @@ private void handleLambdaBasedIterableOperation(IterableContext context, Operati
         return supportedRetTypes;
     }
 
-    private void validateLambdaInputArgs(Operation operation, List<BType> supportedTypes, List<BType> givenTypes) {
-        if (supportedTypes.get(0).tag == TypeTags.ERROR) {
+    private void validateLambdaInputArgs(Operation operation, List<BType> actualRequiredTypes,
+                                         List<BType> userExpectedTypes) {
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IterableAnalyzer.java]: Shall we use the terms `expectedTypes` and the `actualTypes`? Thats the convention we have used everywhere else...

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BIterableTypeVisitor.java [hunk 4]
```diff
@@ -109,14 +109,34 @@ public BIterableTypeVisitor(BLangDiagnosticLog dlog, SymbolTable symTable) {
...
-    protected void logTooMayVariablesError(Operation op) {
+    protected void logTooManyVariablesError(Operation op) {
         dlog.error(op.pos, DiagnosticCode.ITERABLE_TOO_MANY_VARIABLES, op.collectionType);
     }
 
     protected void logNotEnoughVariablesError(Operation op, int count) {
         dlog.error(op.pos, DiagnosticCode.ITERABLE_NOT_ENOUGH_VARIABLES, op.collectionType, count);
     }
 
+    protected BType inferRecordFieldType(BRecordType recordType) {
```
Comment (SupunS) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BIterableTypeVisitor.java]: This method seems to be a duplicate of https://github.com/ballerina-platform/ballerina-lang/pull/10324/files#diff-782094a79e6831ac57a34ad157af49c1R490.

Shall we reuse that?

File: tests/ballerina-unit-test/src/test/resources/test-src/record/open_record_iteration_negative.bal [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        }
+        return false;
+    })
+    .map((string status) => float {
+        if (status == "PASS") {
+            return 4.2;
+        }
+        return 0.0;
+    });
+}
```
Comment (SupunS) [inline tests/ballerina-unit-test/src/test/resources/test-src/record/open_record_iteration_negative.bal]: Add new line

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticCode.java (hunk 1 above)
Comment (pubudu91) [reply compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticCode.java]: Ah this is actually not needed anymore.. Will get rid of this.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 6]
```diff
@@ -476,6 +487,26 @@ public boolean checkStructEquivalency(BType rhsType, BType lhsType) {
         return errorTypes;
     }
 
+    public BType inferRecordFieldType(BRecordType recordType) {
+        boolean isSameType = true;
+        List<BField> fields = recordType.fields;
+        BType inferredType = fields.get(0).type; // If all the fields are the same, doesn't matter which one we pick
+
+        for (int i = 1; i < fields.size(); i++) {
+            isSameType = isSameType && fields.get(i - 1).type.tag == fields.get(i).type.tag;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: do we need the "isSameType" check here? We only need to check if all consecutive fields are of the same type (starting with the 0th index field) or if each field's type is "inferredType" right?

Comment (pubudu91) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java]: True, will change it

**Pull request merged**